trie/hasher.go

```diff
@@ -121,17 +135,44 @@ func (h *hasher) hashChildren(original node, db DatabaseWriter) (node, node, err
 
 	case *fullNode:
 		// Hash the full node's children, caching the newly hashed subtrees
+		var wg sync.WaitGroup
 		collapsed, cached := n.copy(), n.copy()
-
-		for i := 0; i < 16; i++ {
-			if n.Children[i] != nil {
-				collapsed.Children[i], cached.Children[i], err = h.hash(n.Children[i], db, false)
-				if err != nil {
-					return original, original, err
+		if !h.threaded {
+			// Calculate children hash concurrently.
+			// Note. Avoid creating too much goroutines(which may hits GC),
+			// we only thread out on the topmost full node.
+			for i := 0; i < 16; i++ {
+				if n.Children[i] != nil {
+					wg.Add(1)
+					go func(i int) {
+						var herr error
+						collapsed.Children[i], cached.Children[i], herr = h.hash(n.Children[i], db, false)
+						if herr != nil {
+							err = herr
+						}
+						wg.Done()
+					}(i)
+				} else {
+					collapsed.Children[i] = valueNode(nil) // Ensure that nil children are encoded as empty strings.
 				}
-			} else {
-				collapsed.Children[i] = valueNode(nil) // Ensure that nil children are encoded as empty strings.
 			}
+			wg.Wait()
+			h.threaded = true
+		} else {
+			for i := 0; i < 16; i++ {
+				if n.Children[i] != nil {
+					var herr error
+					collapsed.Children[i], cached.Children[i], herr = h.hash(n.Children[i], db, false)
+					if herr != nil {
+						err = herr
+					}
+				} else {
+					collapsed.Children[i] = valueNode(nil) // Ensure that nil children are encoded as empty strings.
+				}
+			}
+		}
+		if err != nil {
+			return original, original, err
```

Comment:
I'm a bit reluctant about this code construct. Seems very brittle that we have the same code duplicated, it's asking for trouble down the line when someone wants to modify it. Couldn't we somehow unify the threaded and non-threaded version?